alacritty_terminal/src/selection.rs

```diff
@@ -165,179 +172,165 @@ impl Selection {
+                {
+                    range.start = prev;
                 }
             }
 
-            // Include all double-width cells and placeholders at bottom right of selection
-            if span.end.line != 0 || span.end.col < num_cols {
-                // Expand from wide char spacer for linewrapping to wide char
-                if (span.end.line + 1 != grid.len() || span.end.col.0 != 0)
-                    && flag_at(span.end, Flags::WIDE_CHAR_SPACER)
-                    && !flag_at(span.end.sub(num_cols.0, 1, true), Flags::WIDE_CHAR)
+            // Expand from wide char to wide char spacer for linewrapping
+            if range.start.line + 1 != grid.len() || range.start.col.0 != 0 {
+                let prev = range.start.sub(num_cols.0, 1, true);
+                if (prev.line + 1 != grid.len() || prev.col.0 != 0)
+                    && flag_at(prev, Flags::WIDE_CHAR_SPACER)
+                    && !flag_at(prev.sub(num_cols.0, 1, true), Flags::WIDE_CHAR)
                 {
-                    span.end = span.end.add(num_cols.0, 1, true);
+                    range.start = prev;
                 }
+            }
+        }
 
-                // Expand from wide char to wide char spacer
-                if flag_at(span.end, Flags::WIDE_CHAR) {
-                    span.end = span.end.add(num_cols.0, 1, true);
-                }
+        // Include all double-width cells and placeholders at bottom right of selection
+        if range.end.line != 0 || range.end.col < num_cols {
+            // Expand from wide char spacer for linewrapping to wide char
+            if (range.end.line + 1 != grid.len() || range.end.col.0 != 0)
+                && flag_at(range.end, Flags::WIDE_CHAR_SPACER)
+                && !flag_at(range.end.sub(num_cols.0, 1, true), Flags::WIDE_CHAR)
+            {
+                range.end = range.end.add(num_cols.0, 1, true);
             }
 
-            span
-        })
+            // Expand from wide char to wide char spacer
+            if flag_at(range.end, Flags::WIDE_CHAR) {
+                range.end = range.end.add(num_cols.0, 1, true);
+            }
+        }
+
+        range
     }
 
     // Bring start and end points in the correct order
-    fn points_need_swap(start: Point<isize>, end: Point<isize>) -> bool {
+    fn points_need_swap(start: Point<usize>, end: Point<usize>) -> bool {
         start.line < end.line || start.line == end.line && start.col > end.col
     }
 
-    // Clamp selection inside the grid to prevent out of bounds errors
+    /// Clamp selection inside grid to prevent OOB.
     fn grid_clamp(
-        mut start: Point<isize>,
-        mut end: Point<isize>,
-        lines: isize,
-        cols: Column,
-    ) -> Option<(Point<isize>, Point<isize>)> {
+        start: &mut Point<usize>,
+        end: &Point<usize>,
+        start_side: &mut Side,
+        is_block: bool,
+        lines: usize,
+    ) -> Result<(), ()> {
```

Comment:
What do you dislike about the in-place mutation? I'd argue in-place mutation is much easier to use than returning values, because if you return things then you're more likely to forget to reassign them properly.

We'd also have to decide if we want to return just start_point an start_side in a tuple, which would be inconsistent because you'd only have to reassign part of the points you give in. Or alternatively, you return start_point, end_point and start_side, which still would be missing end_side but at least returns everything you put in it. However then you have a massive return tuple which is difficult to use.

That's why I personally thing mutating in-place is a better approach here and makes this function much easier to use.

Reply:
because the name is `grid_clamp`, so I'm expecting to get a kind of `clamp` back in a result. So we should either change arguments for this function or its name tbh.

Reply:
Oh, I think we're just understanding things differently then.

I've named the method `grid_clamp`, based on the verb `to clamp`, not the noun `the clamp`. So this method "will clamp the points", not "will return a clamp".

Not sure if this might be lost in translation, maybe our native speaker @nixpulvis can give us a third opinion?

Though if it's misunderstood once, it can happen twice. So I'm open for better suggestions if you have some.

Reply:
I'm mostly relying on this thing here, since it returns clamped value, and not clamps in place. https://en.cppreference.com/w/cpp/algorithm/clamp

Reply:
I've checked once again. Most languages I'm aware of returns something from clamp (c++,rust,swift). So I'm not really sure about a "not returning approach" and using the word `clamp` here, could be confusing. I don't see a problem with returning from this function tbh, even if you use Result<(Anchor,Point),()> or something similar). We can try to name it a bit differently though, but I have no idea here.

Reply:
I couldn't see any way to make this work nicely other than reworking everything in the `to_range` method to use `Anchor`s instead of `Point`s. I think this works quite well since we have to deal with `Side`s anyways, but let me know what you think.